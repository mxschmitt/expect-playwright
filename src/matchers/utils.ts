import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
