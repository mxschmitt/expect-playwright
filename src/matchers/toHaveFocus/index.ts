import { SyncExpectationResult } from "expect/build/types"
import { getMessage, quote } from "../utils"
import { Page } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../../global"

const enum FailureReason {
  NotFound,
  NotFocused,
}

async function isFocused(
  page: Page,
  selector: string,
  options: PageWaitForSelectorOptions = {}
) {
  try {
    await page.waitForSelector(selector, options)
    /* istanbul ignore next */
    const isFocused = await page.$eval(
      selector,
      (el) => el === document.activeElement
    )

    return { pass: isFocused, reason: FailureReason.NotFocused }
  } catch (e) {
    return { pass: false, reason: FailureReason.NotFound }
  }
}

const toHaveFocus: jest.CustomMatcher = async function (
  page: Page,
  selector: string,
  options: PageWaitForSelectorOptions = {}
): Promise<SyncExpectationResult> {
  const result = await isFocused(page, selector, options)

  return {
    pass: result.pass,
    message: () => {
      const not = this.isNot ? " not" : ""
      const hint = this.utils.matcherHint("toHaveFocus", undefined, undefined, {
        isNot: this.isNot,
        promise: this.promise,
      })

      const message =
        result.reason === FailureReason.NotFound
          ? `Expected: element to${not} have focus\n` +
            "Received: element was not found"
          : `Expected: element to${not} have focus`

      return hint + "\n\n" + message
    },
  }
}

export default toHaveFocus
