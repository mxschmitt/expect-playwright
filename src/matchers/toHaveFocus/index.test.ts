import { testWrapper } from "../tests/utils"

import toHaveFocus from "."

describe("toHaveFocus", () => {
  afterEach(async () => {
    await page.setContent("")
  })
  it("positive", async () => {
    await page.setContent(`<input id="foobar">`)
    await page.keyboard.press("Tab")
    const result = await toHaveFocus(page, "#foobar")
    expect(result.pass).toBe(true)
    expect(result.message()).toMatchSnapshot()
  })
  it("negative: target element don't have focus", async () => {
    await page.setContent(`<input id="foo"><input id="bar">`)
    await page.keyboard.press("Tab")
    const result = await toHaveFocus(page, "#bar")
    expect(testWrapper(result)).toThrowErrorMatchingSnapshot()
  })
  it("negative: target element not found", async () => {
    await page.setContent(`<input id="foo">`)
    await page.keyboard.press("Tab")
    const result = await toHaveFocus(page, "#bar", {
      timeout: 1 * 1000,
    })
    expect(testWrapper(result)).toThrowErrorMatchingSnapshot()
  })
  describe("timeout", () => {
    it("positive: should be able to use a custom timeout", async () => {
      setTimeout(async () => {
        await page.setContent(`<input id="foobar">`)
        await page.keyboard.press("Tab")
      }, 500)
      expect(
        testWrapper(
          await toHaveFocus(page, "#foobar", {
            timeout: 1 * 1000,
          })
        )
      ).toBe(true)
    })
    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      expect(
        testWrapper(
          await toHaveFocus(page, "#foobar", {
            timeout: 1 * 1000,
          })
        )
      ).toThrowError()
      const duration = new Date().getTime() - start
      expect(duration).toBeLessThan(1500)
    })
  })
})
