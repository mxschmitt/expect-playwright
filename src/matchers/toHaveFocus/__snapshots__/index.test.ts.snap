// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveFocus negative: target element don't have focus 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoHaveFocus[2m([22m[32mexpected[39m[2m)[22m

Expected: element to have focus"
`;

exports[`toHaveFocus negative: target element not found 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoHaveFocus[2m([22m[32mexpected[39m[2m)[22m

Expected: element to have focus
Received: element was not found"
`;

exports[`toHaveFocus timeout should throw an error after the timeout exceeds 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoHaveFocus[2m([22m[32mexpected[39m[2m)[22m

Expected: element to have focus
Received: element was not found"
`;

exports[`toHaveFocus with 'not' usage negative: target element has focus 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoHaveFocus[2m([22m[32mexpected[39m[2m)[22m

Expected: element to not have focus"
`;
